decayedlog.go

```diff
@@ -0,0 +1,419 @@
+		if err != nil {
+			return fmt.Errorf("Unable to register for epoch "+
+				"notifications: %v", err)
+		}
+
+		d.wg.Add(1)
+		go d.garbageCollector(epochClient)
+	}
+
+	return nil
+}
+
+// initBuckets initializes the primary buckets used by the decayed log, namely
+// the shared hash bucket, and batch replay
+func (d *DecayedLog) initBuckets() error {
+	return d.db.Update(func(tx *bolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(sharedHashBucket)
+		if err != nil {
+			return ErrDecayedLogInit
+		}
+
+		_, err = tx.CreateBucketIfNotExists(batchReplayBucket)
+		if err != nil {
+			return ErrDecayedLogInit
+		}
+
+		return nil
+	})
+}
+
+// Stop halts the garbage collector and closes channeldb.
+func (d *DecayedLog) Stop() error {
+	if !atomic.CompareAndSwapInt32(&d.stopped, 0, 1) {
+		return nil
+	}
+
+	// Stop garbage collector.
+	close(d.quit)
+
+	d.wg.Wait()
+
+	// Close channeldb.
+	d.db.Close()
+
+	return nil
+}
+
+// garbageCollector deletes entries from sharedHashBucket whose expiry height
+// has already past. This function MUST be run as a goroutine.
+func (d *DecayedLog) garbageCollector(epochClient *chainntnfs.BlockEpochEvent) {
+	defer d.wg.Done()
+	defer epochClient.Cancel()
+
+	for {
+		select {
+		case epoch, ok := <-epochClient.Epochs:
+			if !ok {
+				// Block epoch was canceled, shutting down.
+				// TODO(conner): add logging
+				return
+			}
+
+			// Perform a bout of garbage collection using the
+			// epoch's block height.
+			height := uint32(epoch.Height)
+			if err := d.gcExpiredHashes(height); err != nil {
+				// TODO(conner): add logging
+				return
+			}
+
+		case <-d.quit:
+			// Received shutdown request.
+			return
+		}
+	}
+}
+
+// gcExpiredHashes purges the decaying log of all entries whose CLTV expires
+// below the provided height.
+func (d *DecayedLog) gcExpiredHashes(height uint32) error {
+	return d.db.Batch(func(tx *bolt.Tx) error {
+		// Grab the shared hash bucket
+		sharedHashes := tx.Bucket(sharedHashBucket)
+		if sharedHashes == nil {
+			return fmt.Errorf("sharedHashBucket " +
+				"is nil")
+		}
+
+		var expiredCltv [][]byte
+		sharedHashes.ForEach(func(k, v []byte) error {
+			// Deserialize the CLTV value for this entry.
+			cltv := uint32(binary.BigEndian.Uint32(v))
+
+			if cltv < height {
+				// This CLTV is expired. We must add it to an
+				// array which we'll loop over and delete every
+				// hash contained from the db.
+				expiredCltv = append(expiredCltv, k)
+			}
+
+			return nil
+		})
+
+		// Delete every item in the array. This must
+		// be done explicitly outside of the ForEach
+		// function for safety reasons.
+		for _, hash := range expiredCltv {
+			err := sharedHashes.Delete(hash)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+}
+
+// HashSharedSecret Sha-256 hashes the shared secret and returns the first
+// sharedHashSize bytes of the hash.
+func HashSharedSecret(sharedSecret *Hash256) HashPrefix {
```

Comment:
Any particular reason this needs to be publicly exported?

Reply:
can't think of one, _privatizing_